bingo.js

```diff
@@ -598,13 +670,142 @@ function fillVersionSelection()
 	});
 }
 
-function setSquareColor(square, colorClass)
+function setSquareColor(square, colorClass, isTriggeredByUser = false)
 {
 	ALL_COLOURS.forEach(c => square.removeClass(c));
 	square.addClass(colorClass);
+	if (isTriggeredByUser) {
+		updateCurrentSheetProgress();
+	}
 }
 
-function copySeedToClipboard(id, event)
+function loadBingoProgress(sheetToLoad)
+{
+	/* 
+	 * Expected input for sheetToLoad is [1,3,2,5,..]
+	 * where index of array is numeric value of square id
+	 * and cell value is index of ALL_COLOURS 
+	 */
+	if (typeof sheetToLoad !== 'undefined' && sheetToLoad.length) {
+		forEachSquare((i, square) => {
+			$.each(ALL_COLOURS, function(index, colorClass){
+				square.removeClass(colorClass);
+			});
+			square.addClass(ALL_COLOURS[sheetToLoad[i]]);
+		});
+	}
+}
+
+var sheetProgressDebounce = null;
+function updateCurrentSheetProgress()
+{
+	if(sheetProgressDebounce) clearTimeout(sheetProgressDebounce);
+	sheetProgressDebounce = setTimeout(function() {
+			updateCurrentSheetProgressDebounced()
+	}, 100);
+}
+
+function updateCurrentSheetProgressDebounced()
+{
+	if (! isLocalStorageIsAvailable()) {
+		return;
+	}
+
+	processSheetAndStoreAs(getKeyForCurrentProgress());
+}
+
+function processSheetAndStoreAs(localStorageKey)
+{
+	var squares = generateSquareProgress();
+	var haveMarkedSquares = squares.reduce(function(a,b) { return !!a | !!b; });
+	result = {
+		squares: squares,
+		timestamp: Date.now()
+	};
+	
+	// save only if current sheet have marked squares or already in storage
+	if (haveMarkedSquares || localStorage.getItem(localStorageKey)) {
+		pushNewLocalSetting(localStorageKey, JSON.stringify(result));
+	}
+	// update &progress param is such exists
+	if (gup(URL_PARAM_PROGRESS)) {
+		pushNewUrl(generateNewUrlParamWithProgress(squares));
+	}
+	checkAndRemoveOldSaves();
+}
+
+function checkAndRemoveOldSaves()
+{
+	var saves = getSaves();
+	if (saves.length > SHEET_PROGRESS_SAVE_LIMIT) {
+		var oldestSave = saves.pop();
+		if (oldestSave) {
```

Comment:
Function `getSaves` does not put `null`s or `undefined` in its result. Array `saves` has more than `SHEET_PROGRESS_SAVE_LIMIT` completely valid values. This check of `oldestSave` is unnecessary.